sam.c

```diff
@@ -2328,6 +2319,111 @@ int sam_parse1(kstring_t *s, sam_hdr_t *h, bam1_t *b)
     return -2;
 }
 
+static uint32_t read_ncigar(const char *in) {
+    uint32_t n_cigar = 0;
+    char *q = (char *)in;
+    for (; *q && *q != '\t'; ++q)
+        if (!isdigit_c(*q)) ++n_cigar;
+    if (!n_cigar) {
+        hts_log_error("No CIGAR operations");
+        return 0;
+    }
+    if (n_cigar >= 2147483647) {
+        hts_log_error("Too many CIGAR operations");
+        return 0;
+    }
+
+    return n_cigar;
+}
+
+/*! @function
+ @abstract  Parse a CIGAR string into preallocated a uint32_t array
+ @param  in      [in]  pointer to the source string
+ @param  a_cigar [out]  address of the destination uint32_t buffer
+ @return         number of processed input characters; 0 if error
+ */
+static int parse_cigar(const char *in, uint32_t *a_cigar, uint32_t n_cigar) {
+    int i, overflow = 0;
+    char *p, *q = (char *)in;
+    for (i = 0; i < n_cigar; i++) {
+        uint32_t len;
+        int op;
+        p = q;
+        len = hts_str2uint(q, &q, 28, &overflow)<<BAM_CIGAR_SHIFT;
+        if (q == p) {
+            hts_log_error("CIGAR length invalid at position %d (%s)", (int)(i+1), p);
+            return 0;
+        }
+        if (overflow) {
+            hts_log_error("CIGAR length too long at position %d (%.*s)", (int)(i+1), (int)(q-p+1), p);
+            return 0;
+        }
+        op = bam_cigar_table[(unsigned char)*q++];
+        if (op < 0) {
+            hts_log_error("Unrecognized CIGAR operator");
+            return 0;
+        }
+        a_cigar[i] = len;
+        a_cigar[i] |= op;
+    }
+
+    return q-in;
+}
+
+size_t sam_parse_cigar(const char *in, char **end, uint32_t **a_cigar, uint32_t *a_mem, uint32_t *a_len) {
+    size_t n_cigar = 0;
+    int diff;
+
+    if (!in || !end || !a_cigar || !a_mem || !a_len) {
+        hts_log_error("NULL pointer arguments");
+        return 0;
+    }
+    *end = (char *)in;
+
+    n_cigar = read_ncigar(in);
+    if (!n_cigar) return 0;
+    if (n_cigar > (*a_mem - *a_len)) {
+        uint32_t *a_tmp = realloc(*a_cigar, 1.5*(n_cigar + *a_len)*sizeof(**a_cigar));
+        if (a_tmp) {
+            *a_cigar = a_tmp;
+            *a_mem = 1.5*(n_cigar + *a_len);
+        } else {
+            hts_log_error("Memory allocation error");
+            return 0;
+        }
+    }
+
+    if (!(diff = parse_cigar(in, *a_cigar + *a_len, n_cigar))) return 0;
```

Comment:
Why `*a_cigar + *a_len`?

Is this function intented to be ran on multiple CIGAR strings so it can concatenate them?  If so the documentation needs updating.

My assumptions are it would fill out `a_cigar` from pos zero and return `*a_len` as the length filled out, but that's not what it's doing.

Reply:
It offers the flexibility of appending to an existing CIGAR array of `a_len` length. We can drop the feature, if we see no use for it.

Reply:
Flexibility is good, but I'm struggling to find a use case.  Even if we wanted to merge two sequences and their associated CIGAR strings, we'd probably still be wanting to do it manually so we can futz around with merging trailing and leader M ops (for example), or removing hard-clip ops that may end up embedded in the middle otherwise.

So my instinct here is it's not a feature we want.  Others care to weigh in?